teps/0125-add-credential-filter-to-entrypoint-logger.md

```diff
@@ -0,0 +1,337 @@
+tekton pipelines.
+
+### Goals
+
+Secrets are filtered correctly from the output log in a way that the developer 
+of the pipeline does not need to configure this.
+
+Especially output from the `set -x` command should also be filtered correctly.
+
+### Non-Goals
+
+We only target secret values that tekton can automatically redact. We do not want 
+to redact secrets that are dynamically retrieved during the pipeline run and can
+not be known at the start of the pipeline.
+
+We also dont want to redact pipeline parameters containing secret values as part 
+of this TEP.
+
+### Use Cases
+
+When a pipeline is created by the pipeline developer he does not need to configure
+anything. Secrets and credentials used in the pipeline and exposed by logging in the
+pipeline step are automacially redacted from the final log output stream of the tekton
+entrypoint. Also secrets printed via the `set -x` command are redacted from the final
+log output stream.
+
+### Requirements
+
+We do not want to give the entrypoint running in the pipeline step more/any permissions 
+on the API server. 
+It should not call the API server to get secret values for example.
+
+Operators should be able to enable/disable this behaviour via a feature flag in case any
+problems arise from this filtering mechanism.
+
+## Proposal
+
+The tekton controller looks for secret refs in volumes and env and also for CSI volumes 
+referencing secrets in the definition of the pipeline pod. It saves the information in 
+form of the names of environment variables and paths to files with secrets. The pod can
+then read those values and redact them.
+
+This design does not transmit actual secrets and the pod does not need access to the API 
+sever to retrieve them. It tells the pipeline pod only the places where it can find secrets
+attached to it. And those are the secrets that will be redacted. Because only those can be
+used in the corresponding pipeline step.
+
+The credential filter will redact values contained in those secrets from the output log stream
+and replace them with `[REDACTED:<secret-name>]`. Secret name is either the name of the 
+environment variable or the file path where the secret is stored in the pod.
+
+### Notes and Caveats
+
+Secrets can come from different sources than the one implemented in this filter.
+
+Therefore it is important to be very specific about what is being filtered.
+
+
+## Design Details
+
+While creating the pods for the pipeline, the controller will try to detect all secrets 
+attached to the pods. Usually secrets are attached to the pod either by setting environment 
+variables or mounting files into it. These detected secret locations are then transmitted 
+to the pipeline pod for credential filtering.
+
+### Secret Locations
+
+The secret locations detected by the controller are transmitted to the entrypoint as 
+an environment variable in json format with the following format:
```

Comment:
I suspect we may run into limit issues if we stuff it into an environment variable - not sure what the max size is, but I suspect it's fairly small (especially if things like https://devblogs.microsoft.com/oldnewthing/20100203-00/?p=15083 still hold true). You may want to use an annotation + the [downwards API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information) with a file instead. Process is mostly the same, just different fields.

Reply:
SGTM! I think annotations + downwards API are probably a better starting point so we don't pollute the environment for running steps, and we can reserve a path in `/tekton` so there can't be naming collisions. I think the 256kb limit is fine - I don't think this data will be particularly large if it's only path identifiers. We can look into other ways to pass data to the entrypoint if/when this becomes a problem.

Reply:
Ok I adapted the environment variable design to the downward API volume and documented the alternatives we considered.

Reply:
Just realized during implementation that secret locations are container specific. That means we would need one annotation per container, e.g. tekton.dev/secret-locations-<containerName>: "{...}"